### PR TITLE
[InProgress]C_API & DataLayer

### PR DESCRIPTION
**Pull request by chivee**
p

### CODE REVIEW

**Comment by guolinke** [comment]
It has some accuracy drop while changing double to float.
Maybe it is better for our c_api support both for float_32 and float_64.

**Comment by chivee** [comment]
@guolinke , it should be , considering most of functional language mixing 32 and 64 type.

**Comment by Allardvm** [comment]
I see you're making great progress on the API, thanks!

Looking at the various `CreateDataset[...]` functions you expose, I have a suggestion. Since you usually have to process any data you from an external process anyway (binning, etc.), it might be worth exposing two versions of `LGBM_CreateDatasetFromMat` to avoid unnecessary conversion on the external process' end. One `LGBM_CreateDataFromRowMajorMat` for languages that use row-major order arrays, and a second `LGBM_CreateDataFromColMajorMat` for languages that use column-major order arrays. Especially with large databases and (somewhat) limited memory, it could be advantageous to not having to transpose the database array on the external process' end before calling LightGBM.

**Comment by guolinke** [comment]
@Allardvm , It is a good suggestion. I think we can add an parameter like "is_row_major" the avoid expose two functions.

**Comment by Allardvm** [comment]
That sounds great.

**Comment by chivee** [comment]
@guolinke , template will make it easy

**Comment by guolinke** [comment]
@chivee I think we cannot use template for expose c_api. C don't have template...

**Comment by chivee** [comment]
Ops， i forget that


获取 Outlook for Android






On Wed, Nov 2, 2016 at 9:19 PM +0800, "Guolin Ke" <notifications@github.com> wrote:












@chivee I think we cannot use template for expose c_api. C don't have template...

**Comment by guolinke** [comment]
@chivee I have finished the interface for dataset part. you can review or refine it now

File: include/LightGBM/config.h [hunk 1]
```diff
@@ -93,14 +95,15 @@ struct IOConfig: public ConfigBase {
   std::string output_model = "LightGBM_model.txt";
   std::string output_result = "LightGBM_predict_result.txt";
   std::string input_model = "";
-  std::string input_init_score = "";
   int verbosity = 1;
-  int num_model_predict = -1;
+  int num_model_predict = NO_LIMIT;
```
Comment (chivee) [inline include/LightGBM/config.h]: std::numeric_limits<int>::max() will do

File: include/LightGBM/boosting.h [hunk 2]
```diff
@@ -43,28 +43,50 @@ class Boosting {
   virtual void AddDataset(const Dataset* valid_data,
     const std::vector<const Metric*>& valid_metrics) = 0;
 
-  /*! \brief Training logic */
+  /*!
+  * \brief Training logic
+  * \param gradient nullptr for using default objective, otherwise use self-defined boosting
+  * \param hessian nullptr for using default objective, otherwise use self-defined boosting
+  * \param is_eval true if need evalulation or early stop
```
Comment (xuehui1991) [inline include/LightGBM/boosting.h]: evaluation

File: include/LightGBM/c_api.h [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+#define LIGHTGBM_C_API_H_
+#include<cstdint>
+
+
+/*!
+* To avoid type conversion on large data, most of our expose interface support both for float_32 and float_64.
+* Except following:
+* 1. gradients and hessians. 
+* 2. Get current score for training data and validation
+* The reason is becaused they are called frequently, the type-conversion on them maybe time cost.
```
Comment (xuehui1991) [inline include/LightGBM/c_api.h]: because?

File: include/LightGBM/c_api.h [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+* \brief get string message of the last error
+*  all function in this file will return 0 when success
+*  and -1 when an error occured,
+* \return const char* error inforomation
+*/
+DllExport const char* LGBM_GetLastError();
+
+
+// --- start Dataset inferfaces
+
```
Comment (xuehui1991) [inline include/LightGBM/c_api.h]: interface?

File: include/LightGBM/c_api.h [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+  DatesetHandle* out);
+
+/*!
+* \brief create a dataset from CSR format
+* \param indptr pointer to row headers
+* \param indptr_type
+* \param indices findex
+* \param data fvalue
+* \param data_type
+* \param nindptr number of rows in the matix + 1
```
Comment (xuehui1991) [inline include/LightGBM/c_api.h]: matrix?

File: include/LightGBM/c_api.h [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+  DatesetHandle* out);
+
+/*!
+* \brief create a dataset from CSC format
+* \param col_ptr pointer to col headers
+* \param col_ptr_type
+* \param indices findex
+* \param data fvalue
+* \param data_type
+* \param ncol_ptr number of rows in the matix + 1
```
Comment (xuehui1991) [inline include/LightGBM/c_api.h]: matrix?

File: include/LightGBM/c_api.h [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+* \return 0 when success, -1 when failure happens
+*/
+DllExport int LGBM_DatasetGetNumFeature(DatesetHandle handle,
+  int64_t* out);
+
+// --- start Booster interfaces
+
+/*!
+* \brief create an new boosting learner
+* \param train_data traning data set
```
Comment (xuehui1991) [inline include/LightGBM/c_api.h]: training?

File: include/LightGBM/c_api.h [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+* \param is_finished 1 means finised(cannot split any more)
+* \return 0 when success, -1 when failure happens
+*/
+DllExport int LGBM_BoosterUpdateOneIterCustom(BoosterHandle handle,
+  const float* grad,
+  const float* hess,
+  int* is_finished);
+
+/*!
+* \brief get evaluation for training data and validation datas
```
Comment (xuehui1991) [inline include/LightGBM/c_api.h]: * \brief get evaluation for training data and validation data?

File: include/LightGBM/c_api.h [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+* \param out_len len of output result
+* \param out_result used to set a pointer to array
+* \return 0 when success, -1 when failure happens
+*/
+DllExport int LGBM_BoosterGetScore(BoosterHandle handle,
+  int64_t* out_len,
+  const float** out_result);
+
+/*!
+* \brief make prediction for training data and validation datas
```
Comment (xuehui1991) [inline include/LightGBM/c_api.h]: \brief get evaluation for training data and validation data?

File: include/LightGBM/c_api.h [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+
+/*!
+* \brief make prediction for an new data set
+* \param handle handle
+* \param indptr pointer to row headers
+* \param indptr_type 
+* \param indices findex
+* \param data fvalue
+* \param data_type
+* \param nindptr number of rows in the matix + 1
```
Comment (xuehui1991) [inline include/LightGBM/c_api.h]: matrix?

File: include/LightGBM/dataset.h [hunk 11]
```diff
@@ -81,6 +76,13 @@ class Metadata {
   void CheckOrPartition(data_size_t num_all_data,
     const std::vector<data_size_t>& used_data_indices);
 
+
+  void SetLabel(const float* label, data_size_t len);
+
+  void SetWeights(const float* weights, data_size_t len);
+
+  void SetQueryBoundaries(const data_size_t* QueryBoundaries, data_size_t len);
```
Comment (xuehui1991) [inline include/LightGBM/dataset.h]: queryBoundaries?

File: include/LightGBM/dataset.h [hunk 12]
```diff
@@ -237,67 +234,65 @@ class Parser {
...
 
   /*!
   * \brief Save current dataset into binary file, will save to "filename.bin"
   */
-  void SaveBinaryFile();
+  void SaveBinaryFile(const char* bin_filename);
+
+  std::vector<const BinMapper*> GetBinMappers() const;
+
+  void CopyFeatureBinMapperTo(Dataset *dataset, bool is_enable_sparse) const;
```
Comment (xuehui1991) [inline include/LightGBM/dataset.h]: ？

**Comment by chivee** [comment]
@guolinke , it's almost ready to merge to master?

**Comment by chivee** [comment]
refer to #45  #19 #11

**Comment by guolinke** [comment]
@chivee I will merge it after all tests are passed

File: src/boosting/gbdt.cpp [hunk 13]
```diff
@@ -241,17 +250,17 @@ bool GBDT::OutputMetric(int iter) {
...
+            Log::Info("Iteration: %d, %s : %f", iter, name[k].c_str(), test_scores[k]);
+          }
         }
         if (!ret && early_stopping_round_ > 0) {
-          bool the_bigger_the_better = valid_metrics_[i][j]->is_bigger_better();
-          if (best_score_[i][j] < 0
-            || (!the_bigger_the_better && test_scores.back() < best_score_[i][j])
-            || (the_bigger_the_better && test_scores.back() > best_score_[i][j])) {
-            best_score_[i][j] = test_scores.back();
+          auto cur_score = valid_metrics_[i][j]->factor_to_bigger_better() * test_scores.back();
```
Comment (xuehui1991) [inline src/boosting/gbdt.cpp]: ？

File: src/boosting/gbdt.h [hunk 14]
```diff
@@ -42,36 +42,49 @@ class GBDT: public Boosting {
   void AddDataset(const Dataset* valid_data,
        const std::vector<const Metric*>& valid_metrics) override;
   /*!
-  * \brief one training iteration
+  * \brief Training logic
+  * \param gradient nullptr for using default objective, otherwise use self-defined boosting
+  * \param hessian nullptr for using default objective, otherwise use self-defined boosting
+  * \param is_eval true if need evalulation or early stop
```
Comment (xuehui1991) [inline src/boosting/gbdt.h]: evaluation

File: src/c_api.cpp [hunk 15]
```diff
@@ -0,0 +1,841 @@
...
+  bool TrainOneIter(const float* gradients, const float* hessians) {
+    return boosting_->TrainOneIter(gradients, hessians, false);
+  }
+
+  void PrepareForPrediction(int num_used_model, int predict_type) {
+    boosting_->SetNumUsedModel(num_used_model);
+    if (predictor_ != nullptr) { delete predictor_; }
+    bool is_predict_leaf = false;
+    bool is_raw_score = false;
+    if (predict_type == 2) {
```
Comment (guolinke) [inline src/c_api.cpp]: cannot expose enum

Comment (guolinke) [inline src/c_api.cpp]: will update to #define xxx

File: src/c_api.cpp [hunk 16]
```diff
@@ -0,0 +1,841 @@
...
+  int len = 0;
+  boosting->GetPredictAt(data, out_result, &len);
+  *out_len = static_cast<int64_t>(len);
+  return 0;
+}
+
+DllExport int LGBM_BoosterPredictForFile(BoosterHandle handle,
+  int predict_type,
+  int64_t n_used_trees,
+  int data_has_header,
```
Comment (guolinke) [inline src/c_api.cpp]: some languages may not can use bool to pass into interface

File: src/c_api.cpp [hunk 17]
```diff
@@ -0,0 +1,841 @@
...
+      };
+    }
+  } else {
+    Log::Fatal("unknown data type in RowFunctionFromDenseMatric");
+  }
+  return nullptr;
+}
+
+std::function<std::vector<std::pair<int, double>>(int row_idx)>
+RowPairFunctionFromDenseMatric(const void* data, int num_row, int num_col, int data_type, int is_row_major) {
```
Comment (xuehui1991) [inline src/c_api.cpp]: maybe there are more elegant way to combine code between +RowPairFunctionFromDenseMatric  +RowFunctionFromDenseMatric ?

File: src/c_api.cpp [hunk 18]
```diff
@@ -0,0 +1,841 @@
...
+    Log::Fatal("unknown data type in RowFunctionFromCSR");
+  }
+  return nullptr;
+}
+
+std::function<std::vector<std::pair<int, double>>(int idx)>
+ColumnFunctionFromCSC(const void* col_ptr, int col_ptr_type, const int32_t* indices, const void* data, int data_type, int64_t ncol_ptr, int64_t nelem) {
+  if (data_type == dtype_float32) {
+    const float* data_ptr = reinterpret_cast<const float*>(data);
+    if (col_ptr_type == dtype_int32) {
```
Comment (xuehui1991) [inline src/c_api.cpp]: logic in here is too complex..
